instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/HostIdResourceProvider.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.resources;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.ConditionalResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/** {@link ResourceProvider} for automatically configuring <code>host.id</code>. */
+public final class HostIdResourceProvider implements ConditionalResourceProvider {
+
+  private static final Logger logger = Logger.getLogger(HostIdResourceProvider.class.getName());
+
+  public static final AttributeKey<String> HOST_ID = AttributeKey.stringKey("host.id");
+  public static final String REGISTRY_QUERY =
+      "reg query HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Cryptography /v MachineGuid";
+
+  private final Supplier<OsType> getOsType;
+
+  private final Function<Path, List<String>> pathReader;
+
+  private final Supplier<ExecResult> queryWindowsRegistry;
+
+  enum OsType {
+    WINDOWS,
+    LINUX
+  }
+
+  static class ExecResult {
+    int exitCode;
+    List<String> lines;
+
+    public ExecResult(int exitCode, List<String> lines) {
+      this.exitCode = exitCode;
+      this.lines = lines;
+    }
+  }
+
+  public HostIdResourceProvider() {
+    this(
+        HostIdResourceProvider::getOsType,
+        path -> {
+          try {
+            return Files.readAllLines(path);
+          } catch (IOException e) {
+            throw new IllegalStateException(e);
+          }
+        },
+        HostIdResourceProvider::queryWindowsRegistry);
+  }
+
+  // Visible for testing
+  HostIdResourceProvider(
+      Supplier<OsType> getOsType,
+      Function<Path, List<String>> pathReader,
+      Supplier<ExecResult> queryWindowsRegistry) {
+    this.getOsType = getOsType;
+    this.pathReader = pathReader;
+    this.queryWindowsRegistry = queryWindowsRegistry;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    OsType osType = getOsType.get();
```

Comment:
I think the enum is overkill here, especially since you consider everything that's non-windows Linux. I think that approach is fine, but it reduces the two-value enum to a boolean. Recommend removing the enum and instead making a method called `boolean hostIsWindows()`? If it returns true, read the windows way, else try the linux way.

Simpler. Less code. Also avoids the ISE, which is currently impossible. :)

Reply:
The test case passes the osType as an arg - otherwise I'd need to sublass the provider to do that (which is less clear IMO).